Allows for excluding spell id's from the module.

blkht01 opened the pull request:
<!-- First of all, THANK YOU for your contribution. -->

## Changes Proposed:
- Allow to exclude spell IDs from the module

## Issues Addressed:
<!-- If your fix has a relating issue, link it below -->
- Closes

## SOURCE:
<!-- If you can, include a source that can strengthen your claim -->

## Tests Performed:
<!-- Does it build without errors? Did you test in-game? What did you test? On which OS did you test? Describe any other tests performed -->
#### Build and Compatibility:
Compiled successfully without errors on Ubuntu 22.04.
### In-Game Functionality Verification:
#### Module Configuration and Activation:
- Excluded Spell ID 470 in the configuration file.
- Enabled the module for in-game testing.
#### Exclusion Test:
- Executed .learn 470 on an existing character; confirmed the spell was correctly taught.
- Created a new character to verify it did not automatically learn the excluded spell, confirming successful exclusion.
#### Inclusion Test:
- Applied .learn 578 on a character to learn a non-excluded spell.
- Switched between characters to verify the new spell was learned appropriately, confirming the module's selective exclusion works as intended.

#### Outcome:
- The module functions as expected, demonstrating compatibility with the operating system and adherence to configured exclusion rules.

## How to Test the Changes:
<!-- Describe in a detailed step-by-step order how to test the changes -->
1. Update the module's configuration file by adding specific mount spell IDs to the exclusion list.
3. Login with a character and use the .learn command followed by one of the excluded spell IDs to attempt learning the spell.
4. Switch to another character and verify that the previously learned spell is not automatically learned or available, confirming the exclusion works as intended.

blkht01 committed:
Modified to allow for excluding certain spells from the module

blkht01 committed:
change of plans. let's load the spell ids from the config file instead, makes it a bit more versatile for others as well.

blkht01 committed:
Update mod_account_mount.cpp

pangolp commented:
All the changes that are optional, and that can be configured, allow the module to adapt to a greater number of servers. So I think the idea is good. It would be necessary to carry out tests on Windows as well, and see if it passes the tests.

blkht01 committed:
Update .gitignore

ignore .conf file, only commit .dist

blkht01 committed:
Update mod_account_mount.cpp

blkht01 commented:
I don't think the clang fail has anything to do with my changes..? Unless I'm missing something.

I cloned a clean solution and added the module with my commits again to test and it builds fine for me on Ubuntu 22.04.

Edit: I don't have the ability to test on windows unfortunately.

pangolp commented:
> I don't think the clang fail has anything to do with my changes..? Unless I'm missing something.
> 
> I cloned a clean solution and added the module with my commits again to test and it builds fine for me on Ubuntu 22.04.
> 
> Edit: I don't have the ability to test on windows unfortunately.

Once it is finished and has passed all the tests, I will be in charge of testing it in Windows. Although it would be good if you added an example of how to test it and that the list is empty by default. Let each project decide whether or not to add spells to it or leave it empty.

pangolp commented on conf/mod_account_mount.conf.dist:
```diff
@@ -11,3 +11,7 @@ Account.Mounts.Enable = 1
+
+# Excluded Spell IDs (comma-separated)
+
+Account.Mounts.ExcludedSpellIDs = 470,578,6777,66847,17464,18989,10969,10796,35020,34406
```
Leave a line break at the end of the document, so that the error here on Github disappears and by default, the list should be empty to prevent all projects from having these spells added that they may not want to have.

blkht01 committed:
Handle scenarios where no spell IDs are to be excluded

Modified the AccountMounts script to seamlessly handle scenarios where no spell IDs are to be excluded by interpreting "0" or an empty configuration string as no exclusion criteria. This update leverages the existing uint32 type for spell ID representation and enhances the script's flexibility in configuration management.

The change ensures that the AccountMounts script is more adaptable to server administrators' needs, allowing for a straightforward way to toggle spell ID exclusions.

Also added some code commenting

blkht01 commented:
> list is empty by default. Let each project decide whether or not to add spells to it or leave it empty.

Good idea, I added this.

> Although it would be good if you added an example of how to test it

Updated OP

pangolp commented:
Thanks for the changes, and for the modification in the pull request adding new information to better understand and test. When I pass the tests, I will try to review it in Windows, to see how it works. Although I did not know this module and I started working on another one, which works differently. If you want to review it and make modifications, the link is also in the README of the repository.

pangolp reviewed (approved):
I haven't tested the code yet, but the changes I proposed were made. Once the tests have been carried out in the repository, I will test it in Windows to rule out any errors. Thank you for collaborating with the module and the community.

pangolp commented:
```
/home/runner/work/mod-account-mounts/mod-account-mounts/modules/mod-account-mounts/src/mod_account_mount.cpp:18:53: fatal error: 'GetStringDefault' is deprecated: Use GetOption<std::string> instead [-Wdeprecated-declarations]
        std::string excludedSpellsStr = sConfigMgr->GetStringDefault("Account.Mounts.ExcludedSpellIDs", "");
                                                    ^
/home/runner/work/mod-account-mounts/mod-account-mounts/src/common/Configuration/Config.h:57:7: note: 'GetStringDefault' has been explicitly marked deprecated here
    [[deprecated("Use GetOption<std::string> instead")]]
      ^
1 error generated.
make[2]: *** [modules/CMakeFiles/modules.dir/build.make:104: modules/CMakeFiles/modules.dir/mod-account-mounts/src/mod_account_mount.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1199: modules/CMakeFiles/modules.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

pangolp commented on src/mod_account_mount.cpp:
```diff
@@ -2,14 +2,34 @@
+    AccountMounts() : PlayerScript("AccountMounts")
+    {
+        // Retrieve the string of excluded Spell IDs from the config file
+        std::string excludedSpellsStr = sConfigMgr->GetStringDefault("Account.Mounts.ExcludedSpellIDs", "");
```
fatal error: 'GetStringDefault' is deprecated

pangolp replied:
```c++
std::string excludedSpellsStr = sConfigMgr->GetOption<std::string>("Account.Mounts.ExcludedSpellIDs", "");
```

pangolp committed:
fatal error: 'GetStringDefault' is deprecated

pangolp commented:
Thank you for your contribution.
